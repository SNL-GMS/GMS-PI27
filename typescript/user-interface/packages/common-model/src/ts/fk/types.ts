import type { Timeseries } from '../channel-segment/types';
import type { DoubleValue } from '../common/types';
import type {
  ChannelSegmentTypes,
  ChannelTypes,
  FacetedTypes,
  FilterTypes,
  StationTypes,
  WaveformTypes
} from '../common-model';
import type { EntityReference } from '../faceted';

export enum FkUnits {
  FSTAT = 'FSTAT',
  POWER = 'POWER'
}

export interface AzimuthSlownessValues {
  azimuth: number;
  slowness: number;
}

export interface FkMeasuredValues {
  signalDetectionId: string;
  measuredValues: AzimuthSlownessValues;
}
// ***************************************
// Model
// ***************************************
/**
 * Fk meta data from the COI
 */
export interface FkMetadata {
  readonly phaseType: string;
  readonly slowDeltaX: number;
  readonly slowDeltaY: number;
  readonly slowStartX: number;
  readonly slowStartY: number;
}

export interface FstatData {
  readonly azimuthWf: WaveformTypes.Waveform;
  readonly slownessWf: WaveformTypes.Waveform;
  readonly fstatWf: WaveformTypes.Waveform;
}

/**
 * Fk power spectra COI representation
 */
export interface FkPowerSpectraCOI extends Timeseries {
  readonly metadata: FkMetadata;
  readonly values: FkPowerSpectrum[];
  readonly stepSize: number;
  readonly windowLead: number;
  readonly windowLength: number;
  readonly lowFrequency: number;
  readonly highFrequency: number;
}

/**
 * Fk power spectra UI representation
 */
export interface FkPowerSpectra extends FkPowerSpectraCOI {
  // Needed for UI processing added when query returns FkPowerSpectraCOI
  readonly fstatData: FstatData;
  readonly configuration: FkSpectraTemplate;
  readonly reviewed: boolean;
}
export interface FkPowerSpectrum {
  readonly power: number[][];
  readonly fstat: number[][];
  readonly quality: number;
  readonly attributes: OldFkAttributes[];
}

export interface AzimuthSlowness {
  readonly azimuth: number;
  readonly slowness: number;
  readonly azimuthUncertainty: number;
  readonly slownessUncertainty: number;
  readonly extrapolated: boolean;
}
export interface OldFkAttributes extends AzimuthSlowness {
  readonly peakFStat: number;
  readonly xSlow: number;
  readonly ySlow: number;
}

/**
 * FkFrequencyThumbnail preview Fk at a preset FkFrequencyRange
 */
export interface FkFrequencyThumbnail {
  readonly frequencyBand: FkFrequencyRange;
  readonly fkSpectra: FkSpectra;
}

/**
 * Collection of thumbnails by signal detection id
 */
export interface FkFrequencyThumbnailBySDId {
  readonly signalDetectionId: string;
  readonly fkFrequencyThumbnails: FkFrequencyThumbnail[];
}

/**
 * Tracks whether a channel is used to calculate fk
 */

export interface ContributingChannelsConfiguration {
  readonly id: string;
  readonly enabled: boolean;
  readonly name: string;
}

export interface ComputeFkInput {
  readonly startTime: number;
  readonly sampleRate: number;
  readonly sampleCount: number;
  readonly channels: EntityReference<'name'>[];
  readonly windowLead: string;
  readonly windowLength: string;
  readonly lowFrequency: number;
  readonly highFrequency: number;
  readonly useChannelVerticalOffset: boolean;
  readonly phaseType: string;
  readonly normalizeWaveforms: boolean;
  // Optional fields
  readonly slowStartX?: number;
  readonly slowStartY?: number;
  readonly slowDeltaX?: number;
  readonly slowDeltaY?: number;
  readonly slowCountX?: number;
  readonly slowCountY?: number;
}

/**
 * Build FkInput for backend with configuration values to restore
 * with FkSpectra returned in fk configuration
 */

export interface FkInputWithConfiguration {
  readonly fkComputeInput: ComputeFkInput;
  readonly configuration: FkSpectraTemplate;
  readonly signalDetectionId: string;
  readonly isThumbnailRequest: boolean;
}

// ------------------- New FK Definitions Post Pivot ------------------
/** Describes both a FKSpectraWindow and FkSpectrumWindow */
export interface FkWindow {
  duration: number;
  lead: number;
}
export interface FkSpectraMetadata {
  readonly fkSpectrumWindow: FkWindow;
  readonly slownessGrid: SlownessGrid;
  readonly phase: string;
}
/**
 * Fk spectra COI representation
 */
export interface FkSpectraCOI extends Timeseries {
  readonly fkSpectraMetadata?: FkSpectraMetadata;
  readonly samples: FkSpectrum[];
  readonly peakFkAttributes?: FkAttributes; // TODO take away optional
}
/**
 * Fk spectra with UI representation
 */
export interface FkSpectra extends FkSpectraCOI {
  // Needed for UI processing added when query returns FkPowerSpectraCOI
  readonly configuration: FkSpectraTemplate;
  readonly reviewed: boolean;
  readonly fstatData: FstatData;
}

export interface FkSpectrum {
  readonly power: number[][];
  readonly fstat: number[][];
  readonly fkQual?: number;
  readonly fkAttributes?: FkAttributes[];
}
export interface FkAttributes {
  readonly peakFStat: number;
  readonly receiverToSourceAzimuth: DoubleValue;
  readonly slowness: DoubleValue;
}

/**
 * FK Spectra Template Definition Components
 */
export interface SlownessGrid {
  maxSlowness: number;
  numPoints: number;
}
export interface FkWaveformSampleRate {
  waveformSampleRateHz: number;
  waveformSampleRateToleranceHz: number;
}
export interface FkFrequencyRange {
  lowFrequencyHz: number;
  highFrequencyHz: number;
}
export interface FkFrequencyRangeWithPrefilter extends FkFrequencyRange {
  previewPreFilterDefinition:
    | FilterTypes.FilterDefinition
    | FilterTypes.LinearFilterDefinition
    | FilterTypes.CascadeFilterDefinition
    | FilterTypes.PhaseMatchFilterDefinition
    | FilterTypes.AutoRegressiveFilterDefinition;
}
export enum FkUncertaintyOption {
  EMPIRICAL = 'EMPIRICAL',
  EXPONENTIAL_SIGNAL_COHERENCE = 'EXPONENTIAL_SIGNAL_COHERENCE',
  OBSERVED_SIGNAL_COHERENCE = 'OBSERVED_SIGNAL_COHERENCE',
  PERFECT_SIGNAL_COHERENCE = 'PERFECT_SIGNAL_COHERENCE'
}
export enum TaperFunction {
  BLACKMAN = 'BLACKMAN',
  COSINE = 'COSINE',
  HAMMING = 'HAMMING',
  HANNING = 'HANNING',
  PARZEN = 'PARZEN',
  WELCH = 'WELCH'
}

export interface ProcessingMasksByChannel {
  channel: ChannelTypes.Channel;
  processingMasks: ChannelSegmentTypes.ProcessingMask[];
}

export interface FkSpectraParameters {
  phase: string;
  preFilter: FilterTypes.FilterDefinition | undefined;
  slownessGrid: SlownessGrid;
  fkSpectrumWindow: FkWindow;
  fkFrequencyRange: FkFrequencyRange;
  waveformSampleRate: FkWaveformSampleRate;
  spectrumStepDuration: number;
  orientationAngleToleranceDeg: number;
  minimumWaveformsForSpectra: number;
  normalizeWaveforms: boolean;
  twoDimensional: boolean;
  fftTaperFunction: TaperFunction;
  fkUncertaintyOption: FkUncertaintyOption;
  fftTaperPercent: number;
}
export interface FkSpectraDefinition {
  orientationAngles: ChannelTypes.OrientationAngles;
  fkParameters: FkSpectraParameters;
}
export interface FkSpectraTemplate {
  inputChannels: FacetedTypes.EntityReference<'name', ChannelTypes.Channel>[];
  fkSpectraParameters: FkSpectraParameters;
  fkSpectraWindow: FkWindow;
  station: FacetedTypes.VersionReference<'name', StationTypes.Station>;
  phaseType: string;
}

export type FkSpectraTemplatesByStationByPhase = Record<
  string /* station name */,
  Record<string /* phase type */, FkSpectraTemplate>
>;

export type FkReviewablePhasesByStation = Record<string, string[]>;

export type FkReviewablePhasesByActivityNameByStation = Record<string, FkReviewablePhasesByStation>;
