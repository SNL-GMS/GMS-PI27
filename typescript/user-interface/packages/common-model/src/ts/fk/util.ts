import type { ChannelSegment, Timeseries } from '../channel-segment/types';
import { TimeseriesType } from '../channel-segment/types';
import type { FkTypes, SignalDetectionTypes } from '../common-model';
import { findPhaseFeatureMeasurementValue, getCurrentHypothesis } from '../signal-detection/util';
import type { FkFrequencyRange, FkPowerSpectra, FkSpectra } from './types';

/** Kilometer per degree of Earth */
const KM_PER_DEGREE = 111.19492664;

/**
 * Checks if FK spectra channel segment
 *
 * @param object Channel Segment
 * @returns boolean
 */
export function isFkSpectraChannelSegment(
  object: ChannelSegment<Timeseries>
): object is ChannelSegment<FkSpectra> {
  return object.timeseriesType === TimeseriesType.FK_SPECTRA;
}

/**
 * Checks if FK spectra
 *
 * @param object FkSpectra Timeseries
 * @returns boolean
 */
export function isFkSpectra(object: Timeseries): object is FkSpectra {
  return object.type === TimeseriesType.FK_SPECTRA;
}

/**
 * Checks if FK power spectra channel segment
 *
 * @param object Channel Segment
 * @returns boolean
 */
export function isFkPowerSpectraChannelSegment(
  object: ChannelSegment<Timeseries>
): object is ChannelSegment<FkSpectra> {
  return object.timeseriesType === TimeseriesType.FK_SPECTRA_OLD;
}

/**
 * Checks if FK spectra
 *
 * @param object FkSpectra Timeseries
 * @returns boolean
 */
export function isFkPowerSpectra(object: Timeseries): object is FkPowerSpectra {
  return object.type === TimeseriesType.FK_SPECTRA_OLD;
}

// TODO: Review to see what might make more sense as part of configuration
/** Width/height of y and x axis respectively */
export const SIZE_OF_FK_RENDERING_AXIS_PX = 35;

/** The height of the rendering needs to be 8 pixels smaller than the width */
export const FK_RENDERING_HEIGHT_OFFSET = 8;
/** The width of the x axis markers needs to be 10 pixels wider than the canvas */
export const FK_X_AXIS_WIDTH_OFFSET = 10;
/** The height of the y axis markers needs to be 12 pixels taller than the canvas */
export const FK_Y_AXIS_HEIGHT_OFFSET = 12;

export const BEAM_TOOLBAR_WIDTH_OFFSET = -80;

/** Default Frequency Bands */
export const FrequencyBands: FkFrequencyRange[] = [
  {
    lowFrequencyHz: 0.5,
    highFrequencyHz: 2
  },
  {
    lowFrequencyHz: 1,
    highFrequencyHz: 2.5
  },
  {
    lowFrequencyHz: 1.5,
    highFrequencyHz: 3
  },
  {
    lowFrequencyHz: 2,
    highFrequencyHz: 4
  },
  {
    lowFrequencyHz: 3,
    highFrequencyHz: 6
  }
];

/**
 * Formats a frequency band into a string for the drop down
 *
 * @param band Frequency band to format
 */
export function frequencyBandToString(band: FkTypes.FkFrequencyRange): string {
  return `${band.lowFrequencyHz} - ${band.highFrequencyHz} Hz`;
}

/**
 * Creates menu options for frequency bands
 */
export const generateFrequencyBandOptions = (): string[] => {
  const items: string[] = [];
  FrequencyBands.forEach(frequency => {
    items.push(frequencyBandToString(frequency));
  });
  return items;
};

/**
 * Approximate conversion between degrees and km
 *
 * @returns Kilometer
 */
export function degreeToKmApproximate(degree: number): number {
  const DEGREES_IN_CIRCLE = 360;
  const RAD_EARTH = 6371;
  const TWO_PI = Math.PI * 2;
  return degree / (DEGREES_IN_CIRCLE / (RAD_EARTH * TWO_PI));
}

/**
 * Approximate conversion between km and degrees
 *
 * @returns Degree
 */
export function kmToDegreesApproximate(km: number): number {
  const DEGREES_IN_CIRCLE = 360;
  const RAD_EARTH = 6371;
  const TWO_PI = Math.PI * 2;
  return km * (DEGREES_IN_CIRCLE / (RAD_EARTH * TWO_PI));
}

/**
 * @returns Input value in s/degree converted to km/s
 */
export function convertSecondsPerDegreeToKilometersPerSecond(value: number) {
  return KM_PER_DEGREE / value;
}

/**
 * @returns Input value in s/degree converted to s/km
 */
export function convertSecondsPerDegreeToSecondsPerKilometer(value: number) {
  return 1 / convertSecondsPerDegreeToKilometersPerSecond(value);
}

/**
 * Determines if the fk needs review
 *
 * @param sd: the signal detection to check
 * @returns true if needs review, otherwise false
 */
export const fkNeedsReview = (
  fk: FkTypes.FkSpectra | undefined,
  phasesNeedingReview: string[] | undefined,
  sd: SignalDetectionTypes.SignalDetection | undefined,
  isForFiltering = false
): boolean => {
  if (!fk || !sd || !phasesNeedingReview) {
    return false;
  }
  if (fk.reviewed && !isForFiltering) {
    return false;
  }
  const currentSdHypothesis = getCurrentHypothesis(sd.signalDetectionHypotheses);
  const phase = findPhaseFeatureMeasurementValue(currentSdHypothesis.featureMeasurements)?.value;

  if (phase && phasesNeedingReview.indexOf(phase) > -1) {
    return true;
  }
  return false;
};
