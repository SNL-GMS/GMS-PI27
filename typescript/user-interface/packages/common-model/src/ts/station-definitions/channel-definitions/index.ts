export * from './channel-definitions';
export * as Util from './util';
