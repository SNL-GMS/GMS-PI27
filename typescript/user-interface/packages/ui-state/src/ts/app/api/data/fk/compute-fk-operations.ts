import type {
  ChannelSegmentTypes,
  ChannelTypes,
  FkTypes,
  ProcessingMaskDefinitionTypes,
  StationTypes,
  WaveformTypes
} from '@gms/common-model';

import type { UiChannelSegment } from '../../../../types';
import { computeFkApi, getPeakFkAttributesApi } from '../../../../workers/api';
import type { ComputeFkSpectraArgs } from './types';

interface ComputeFkParams {
  fkSpectraDefinition: FkTypes.FkSpectraDefinition;
  station: StationTypes.Station;
  inputChannels: ChannelTypes.Channel[];
  detectionTime: number;
  startTime: number;
  endTime: number;
  uiChannelSegments: UiChannelSegment<WaveformTypes.Waveform>[];
  processingMasksByChannel: {
    channel: ChannelTypes.Channel;
    processingMasks: ChannelSegmentTypes.ProcessingMask[];
  }[];
  maskTaperDefinition: ProcessingMaskDefinitionTypes.TaperDefinition | undefined;
  expandedTimeBufferSeconds: number;
}

export async function computeFk(
  params: ComputeFkParams
): Promise<ChannelSegmentTypes.ChannelSegment<FkTypes.FkSpectraCOI>> {
  const {
    fkSpectraDefinition,
    detectionTime,
    startTime,
    endTime,
    station,
    inputChannels,
    uiChannelSegments,
    processingMasksByChannel,
    maskTaperDefinition,
    expandedTimeBufferSeconds
  } = params;

  const computeFkSpectraArgs: ComputeFkSpectraArgs = {
    fkSpectraDefinition,
    station,
    inputChannels,
    detectionTime,
    startTime,
    endTime,
    uiChannelSegments,
    processingMasksByChannel,
    maskTaperDefinition,
    expandedTimeBufferSeconds
  };

  // Call FK worker
  return computeFkApi(computeFkSpectraArgs);
}

export async function getPeakFkAttributes(
  fkSpectra: FkTypes.FkSpectraCOI
): Promise<FkTypes.FkAttributes> {
  // Call FK worker
  return getPeakFkAttributesApi(fkSpectra);
}
