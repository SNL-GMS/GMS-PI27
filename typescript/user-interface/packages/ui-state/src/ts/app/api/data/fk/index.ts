export * from './compute-legacy-fk-spectra';
export * from './fk-utils';
export * from './mark-fk-reviewed';
export * from './mutate-fk-channel-segment-record';
export * from './types';
