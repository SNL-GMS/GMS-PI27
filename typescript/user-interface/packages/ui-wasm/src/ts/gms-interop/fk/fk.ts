import type {
  ChannelSegmentTypes,
  ChannelTypes,
  CommonTypes,
  FkTypes,
  ProcessingMaskDefinitionTypes,
  StationTypes,
  WaveformTypes
} from '@gms/common-model';
import { UILogger } from '@gms/ui-util';

import type { GmsInteropModule, Wasm } from '../gms-interop-module';
import { getInteropModule } from '../gms-interop-module';
import {
  convertToWasmChannelSegments,
  convertToWasmFkSpectra,
  convertToWasmFkSpectraDefinition,
  convertToWasmProcessingMasksByChannelMap,
  convertToWasmStation,
  convertToWasmTaperDefinition
} from '../ts-to-wasm-converters';
import {
  convertToTsFkAttributes,
  convertToTsFkTimeseriesWithMissingInputChannels
} from '../wasm-to-ts-converters';

const logger = UILogger.create('GMS_COMPUTE_FK', process.env.GMS_COMPUTE_FK);

/** Prop object for {@link ComputeFkArgs} */
export interface ComputeFkArgs {
  readonly fkSpectraDefinition: FkTypes.FkSpectraDefinition;
  readonly station: StationTypes.Station;
  readonly inputChannelNames: string[];
  readonly detectionTime: number;
  readonly startTime: number;
  readonly endTime: number;
  readonly channelSegments: ChannelSegmentTypes.ChannelSegment<WaveformTypes.Waveform>[];
  readonly processingMasksByChannel: Record<
    ChannelTypes.Channel['name'],
    ChannelSegmentTypes.ProcessingMask[]
  >;
  readonly maskTaperDefinition: ProcessingMaskDefinitionTypes.TaperDefinition | undefined;
}

/**
 * Performs the computeFk operation.
 *
 * ! Invokes the WASM algorithm
 *
 */
export async function computeFkWasm(
  props: ComputeFkArgs
): Promise<CommonTypes.TimeseriesWithMissingInputChannels<FkTypes.FkSpectraCOI>> {
  let gmsInteropModule: GmsInteropModule | null = await getInteropModule().catch(() => {
    throw new Error(`Failed access the interop module.`);
  });

  const fkComputeProvider: Wasm.FkComputeUtility = new gmsInteropModule.FkComputeUtility();

  let fkSpectraDefinition: Wasm.FkSpectraDefinition | undefined;
  let station: Wasm.Station | undefined;
  const inputChannelNames: Wasm.VectorString = new gmsInteropModule.VectorString();
  let processingMasks: Wasm.ProcessingMasksByChannelMap | undefined;
  let taperDefinition: Wasm.TaperDefinition | undefined;
  let channelSegments: Wasm.VectorChannelSegment | undefined;

  try {
    fkSpectraDefinition = convertToWasmFkSpectraDefinition(
      gmsInteropModule,
      props.fkSpectraDefinition
    );
    station = convertToWasmStation(gmsInteropModule, props.station);
    props.inputChannelNames.forEach(channelName => inputChannelNames.push_back(channelName));
    processingMasks = convertToWasmProcessingMasksByChannelMap(
      gmsInteropModule,
      props.processingMasksByChannel
    );
    channelSegments = convertToWasmChannelSegments(gmsInteropModule, props.channelSegments);
    taperDefinition = props.maskTaperDefinition
      ? convertToWasmTaperDefinition(gmsInteropModule, props.maskTaperDefinition)
      : props.maskTaperDefinition;

    const result = convertToTsFkTimeseriesWithMissingInputChannels(
      gmsInteropModule,
      fkComputeProvider.computeFk(
        fkSpectraDefinition,
        station,
        inputChannelNames,
        props.detectionTime,
        props.startTime,
        props.endTime,
        channelSegments,
        processingMasks,
        taperDefinition
      )
    );
    logger.info(`Computed FK`, JSON.stringify(result));
    return result;
  } catch (error) {
    try {
      const [type, message] = gmsInteropModule.getExceptionMessage(error);
      logger.error(`Failed to compute fk`, { type, error });
      // eslint-disable-next-line no-underscore-dangle
      gmsInteropModule._free(error);
      throw new Error(`${type}: ${message}`);
    } catch (runtimeError) {
      logger.error(`Failed to compute fk`, {
        error
      });
      throw runtimeError;
    }
  } finally {
    fkComputeProvider.delete();
    if (fkSpectraDefinition) fkSpectraDefinition.delete();
    if (station) station.delete();
    if (inputChannelNames) inputChannelNames.delete();
    if (processingMasks) processingMasks.delete();
    if (taperDefinition) taperDefinition.delete();
    if (channelSegments) channelSegments.delete();
    gmsInteropModule = null;
  }
}

export async function getPeakFkAttributesWasm(
  fkSpectraTs: FkTypes.FkSpectraCOI
): Promise<FkTypes.FkAttributes> {
  let gmsInteropModule: GmsInteropModule | null = await getInteropModule().catch(() => {
    throw new Error(`Failed access the interop module.`);
  });

  const fkComputeProvider = new gmsInteropModule.FkComputeUtility();
  let wasmResult: Wasm.FkAttributes | undefined;
  let fkSpectra: Wasm.FkSpectra | undefined;

  try {
    fkSpectra = convertToWasmFkSpectra(gmsInteropModule, fkSpectraTs);
    wasmResult = fkComputeProvider.getPeakFkAttributes(fkSpectra);
    const result = convertToTsFkAttributes(gmsInteropModule, wasmResult);
    logger.info(`Peak attributes`, result);
    return result;
  } catch (error) {
    logger.error(`Failed get peak attributes`, error);
    throw error;
  } finally {
    fkComputeProvider.delete();
    if (fkSpectra) fkSpectra.delete();
    if (wasmResult) wasmResult.delete();
    gmsInteropModule = null;
  }
}
