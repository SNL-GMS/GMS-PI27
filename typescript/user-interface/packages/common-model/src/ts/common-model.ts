import * as ArrayUtil from './array-util';
import * as BeamformingTemplateTypes from './beamforming-templates/types';
import * as ChannelSegmentTypes from './channel-segment';
import * as ColorTypes from './color/types';
import * as CommonTypes from './common';
import * as Displays from './displays/types';
import * as Endpoints from './endpoints/types';
import * as EventTypes from './event';
// required for legacy components
import * as LegacyEventTypes from './event/legacy';
import * as FacetedTypes from './faceted';
import * as FilterUtil from './filter/filter-util';
import * as FilterTypes from './filter/types';
import * as FkTypes from './fk';
import * as ProcessingMaskDefinitionTypes from './processing-mask-definitions/types';
import * as QcSegmentTypes from './qc-segment';
import * as RotationTypes from './rotation/types';
import * as SignalDetectionTypes from './signal-detection';
import * as ChannelTypes from './station-definitions/channel-definitions';
import * as ResponseTypes from './station-definitions/response-definitions';
import * as StationTypes from './station-definitions/station-definitions/station-definitions';
import * as ProcessingStationTypes from './station-processing/types';
import * as ReferenceStationTypes from './station-reference/types';
import * as TimeTypes from './time';
import * as TypeUtil from './type-util/type-util';
import * as ConfigurationTypes from './ui-configuration/types';
import * as UserProfileTypes from './user-profile/types';
import * as WaveformTypes from './waveform/types';
import * as WaveformUtil from './waveform/util';
import * as WorkflowTypes from './workflow/types';

export * from './faceted/utils';

export {
  ArrayUtil,
  BeamformingTemplateTypes,
  ChannelSegmentTypes,
  ChannelTypes,
  ColorTypes,
  CommonTypes,
  ConfigurationTypes,
  Displays,
  Endpoints,
  EventTypes,
  FacetedTypes,
  FilterTypes,
  FilterUtil,
  FkTypes,
  LegacyEventTypes,
  ProcessingMaskDefinitionTypes,
  ProcessingStationTypes,
  QcSegmentTypes,
  ReferenceStationTypes,
  ResponseTypes,
  RotationTypes,
  SignalDetectionTypes,
  StationTypes,
  TimeTypes,
  TypeUtil,
  UserProfileTypes,
  WaveformTypes,
  WaveformUtil,
  WorkflowTypes
};
