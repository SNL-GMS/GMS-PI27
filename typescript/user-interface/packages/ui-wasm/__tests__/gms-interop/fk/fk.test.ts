import type { ChannelTypes, StationTypes, WaveformTypes } from '@gms/common-model';
import { ChannelSegmentTypes, FkTypes, QcSegmentTypes } from '@gms/common-model';
import { PD01Channel, pdar } from '@gms/common-model/__tests__/__data__/station-definitions';

import type { ComputeFkArgs } from '../../../src/ts/gms-interop/fk/fk';
import { computeFkWasm, getPeakFkAttributesWasm } from '../../../src/ts/gms-interop/fk/fk';
import { defaultSpectraDefinition, fkData } from '../test-data/fk-data';
import { genericTestData } from '../test-data/generic-test-data-3s';

const pdChannelName = 'PDAR.PD01.SHZ';
const qcSegmentVersion: QcSegmentTypes.QcSegmentVersion = {
  id: { parentQcSegmentId: 'qcSegmentTestId', effectiveAt: 0 },
  startTime: 0,
  endTime: 100,
  createdBy: 'User 1',
  rejected: false,
  rationale: '',
  type: QcSegmentTypes.QcSegmentType.CALIBRATION,
  discoveredOn: undefined,
  stageId: undefined,
  category: QcSegmentTypes.QcSegmentCategory.ANALYST_DEFINED,
  channels: [{ name: pdChannelName, effectiveAt: 0 }]
};
const pd01ProcessingMask: ChannelSegmentTypes.ProcessingMask = {
  id: 'processing mask',
  effectiveAt: 0,
  startTime: 100,
  endTime: 200,
  appliedToRawChannel: { name: pdChannelName },
  processingOperation: ChannelSegmentTypes.ProcessingOperation.AMPLITUDE_MEASUREMENT_BEAM,
  maskedQcSegmentVersions: [qcSegmentVersion]
};

const pd01ProcessingMaskRecord: Record<string, ChannelSegmentTypes.ProcessingMask[]> = {
  [pdChannelName]: [pd01ProcessingMask]
};

const relativePositionByChannelMap: Record<string, ChannelTypes.RelativePosition> =
  pdar.relativePositionsByChannel;

describe('ComputeFk', () => {
  /**
   * TODO: This test is handing the wrong channel segments for PDAR, they are coming in from
   * std::invalid_argument: Failed to retrieve the relative position for channel name MA2.MA2.BHE
   */
  test('interop computeFkWasm', async () => {
    const chanSegs: ChannelSegmentTypes.ChannelSegment<WaveformTypes.Waveform>[] = [];
    const chanSeg: ChannelSegmentTypes.ChannelSegment<WaveformTypes.Waveform> = {
      ...genericTestData[0],
      id: {
        ...genericTestData[0].id,
        channel: {
          name: PD01Channel.name,
          effectiveAt: PD01Channel.effectiveAt
        }
      }
    };
    chanSegs.push(chanSeg);

    const props: ComputeFkArgs = {
      fkSpectraDefinition: defaultSpectraDefinition,
      station: {
        name: 'PDAR',
        effectiveAt: 100,
        relativePositionsByChannel: relativePositionByChannelMap
      } as StationTypes.Station,
      inputChannelNames: [pdChannelName],
      detectionTime: 200,
      startTime: 100,
      endTime: 310,
      channelSegments: chanSegs,
      processingMasksByChannel: pd01ProcessingMaskRecord,
      maskTaperDefinition: {
        taperLengthSamples: 100,
        taperFunction: FkTypes.TaperFunction.COSINE
      }
    };

    const result = await computeFkWasm(props);
    expect(result).toBeDefined();
    expect(result.missingInputChannels).toHaveLength(1);
    expect(result.missingInputChannels).toMatchSnapshot();
    expect(result.timeseries).toHaveLength(1);
    expect(result.timeseries[0].startTime).toEqual(props.startTime);
    expect(result.timeseries[0].endTime).toEqual(props.endTime);
    expect(result.timeseries[0]).toMatchSnapshot();
  });
});

describe('GetPeakAttributes', () => {
  test('interop getPeakFkAttributesWasm', async () => {
    const result = await getPeakFkAttributesWasm(fkData);
    expect(result).toBeDefined();
    expect(result.peakFStat).not.toBeUndefined();
    expect(result.receiverToSourceAzimuth).not.toBeUndefined();
    expect(result.slowness).not.toBeUndefined();
    expect(result).toMatchSnapshot();
  });
});
