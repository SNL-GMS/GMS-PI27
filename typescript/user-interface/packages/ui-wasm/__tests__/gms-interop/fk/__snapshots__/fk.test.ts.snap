// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComputeFk interop computeFkWasm 1`] = `
[
  {
    "channel": {
      "name": "PDAR.PD01.SHZ",
    },
    "timeRanges": [
      {
        "endTime": 310,
        "startTime": 100,
      },
    ],
  },
]
`;

exports[`ComputeFk interop computeFkWasm 2`] = `
{
  "endTime": 310,
  "fkSpectraMetadata": {
    "fkSpectrumWindow": {
      "duration": 5.1,
      "lead": 1.1,
    },
    "phase": "P",
    "slownessGrid": {
      "maxSlowness": 40,
      "numPoints": 81,
    },
  },
  "sampleCount": 1,
  "sampleRateHz": 20,
  "samples": [],
  "startTime": 100,
  "type": "FK_SPECTRA",
}
`;

exports[`GetPeakAttributes interop getPeakFkAttributesWasm 1`] = `
{
  "peakFStat": 5,
  "receiverToSourceAzimuth": {
    "standardDeviation": 1,
    "units": "DECIBELS",
    "value": 2,
  },
  "slowness": {
    "standardDeviation": 1,
    "units": "DECIBELS",
    "value": 2,
  },
}
`;
